### PR TITLE
Integrate metabolomics reid logic to re_id_tool.py

### PR DESCRIPTION
**Pull request by mbthornton-lbl**
This PR provides an update to the re_id_tool.py command `update-study`

Updates:

- Omics record:  update `has_output`
- Metab.has_input data objects
- Metab.has_output data objects
- Metab.has_calibration data objects

Same updates for NOM, except no calibration data

### CODE REVIEW

File: nmdc_automation/re_iding/scripts/re_id_tool.py [hunk 1]
```diff
@@ -92,6 +91,188 @@ def cli(ctx, target):
...
+    new_output_data_object_ids = []
+    for output_data_object in output_data_objects:
+        # mint new ID for the data object
+        legacy_id = output_data_object["id"]
+        new_id = api_client.minter("nmdc:DataObject")
+        new_output_data_object_ids.append(new_id)
+        updated_record_identifiers.append(("data_object_set", legacy_id, new_id))
+        # update the data object record
+        do_filter_criteria = {"id": legacy_id}
+        do_update_criteria = {"$set": {"id": new_id, "alternative_identifiers": [legacy_id], "generated_by": metabolomics_new_id}}
```
Comment (aclum) [inline nmdc_automation/re_iding/scripts/re_id_tool.py:271]: slot is 'was_generated_by' not 'generated_by'

File: nmdc_automation/re_iding/scripts/re_id_tool.py [hunk 2]
```diff
@@ -92,6 +91,188 @@ def cli(ctx, target):
...
+
+        # Find Metabolomics Activity records that are informed_by the legacy OmicsProcessing ID
+        metabolomics_activity_records = db_client["metabolomics_analysis_activity_set"].find(
+            {"was_informed_by": omics_legacy_id}
+        )
+        len_metabolomics_activity_records = len(list(metabolomics_activity_records.clone()))
+        logging.info(f"Found {len_metabolomics_activity_records} Metabolomics Activity records for {omics_legacy_id}")
+
+
+        for metabolomics_activity_record in metabolomics_activity_records:
```
Comment (aclum) [inline nmdc_automation/re_iding/scripts/re_id_tool.py:144]: metabolomics analysis activity records also need their has_calibration values updated. If the value is 'false' or 'False' do nothing, if the value has an emsl prefix, find that record in the data object collection, update that record and update the metabolomics analysis activity record.

File: nmdc_automation/re_iding/scripts/re_id_tool.py [hunk 3]
```diff
@@ -92,6 +91,188 @@ def cli(ctx, target):
...
+                        logging.info(f"Updated {result.modified_count} omics_processing_set records")
+
+                        # update the has_output data objects and the Metabolomics Activity record
+                        metabolomics_new_id = api_client.minter("nmdc:MetabolomicsAnalysisActivity")
+                        non_orphan_metabolomics_ids.append(metabolomics_new_id)
+                        updated_record_identifiers.append(("metabolomics_analysis_activity_set", metabolomics_legacy_id, metabolomics_new_id))
+                        new_output_data_object_ids = _update_metabolomics_output_data_objects(
+                            output_data_objects, metabolomics_new_id, db_client, api_client, updated_record_identifiers
+                            )
+                        # update the Metabolomics Activity record
```
Comment (aclum) [inline nmdc_automation/re_iding/scripts/re_id_tool.py:176]: This block is missing updating has_input for the metabolomics activity record.

**Review by aclum — CHANGES_REQUESTED** [review]
See individual comments about was_generated_by for data_object_set records, has_calibration for  and has_input for metabolomics_analysis_activity_set records.

File: nmdc_automation/re_iding/scripts/re_id_tool.py (hunk 1 above)
Comment (mbthornton-lbl) [reply nmdc_automation/re_iding/scripts/re_id_tool.py]: https://github.com/microbiomedata/nmdc_automation/pull/116/commits/471705a6b14df41909af558e829d7c6acc52327e

File: nmdc_automation/re_iding/scripts/re_id_tool.py (hunk 2 above)
Comment (mbthornton-lbl) [reply nmdc_automation/re_iding/scripts/re_id_tool.py:144]: https://github.com/microbiomedata/nmdc_automation/pull/116/commits/4bf4eb05c60ea1aad1b12501c8ec09ed1e42aa56

File: nmdc_automation/re_iding/scripts/re_id_tool.py (hunk 3 above)
Comment (mbthornton-lbl) [reply nmdc_automation/re_iding/scripts/re_id_tool.py]: https://github.com/microbiomedata/nmdc_automation/pull/116/commits/5066cbf06d9156a0b0b598c246e4f43a6ed0ad2b

File: nmdc_automation/re_iding/scripts/re_id_tool.py [hunk 4]
```diff
@@ -92,6 +91,219 @@ def cli(ctx, target):
...
+    new_output_data_object_ids = []
+    for output_data_object in output_data_objects:
+        # mint new ID for the data object
+        legacy_id = output_data_object["id"]
+        new_id = api_client.minter("nmdc:DataObject")
+        new_output_data_object_ids.append(new_id)
+        updated_record_identifiers.append(("data_object_set", legacy_id, new_id))
+        # update the data object record
+        do_filter_criteria = {"id": legacy_id}
+        do_update_criteria = {"$set": {"id": new_id, "alternative_identifiers": [legacy_id], "was_generated_by":
```
Comment (aclum) [inline nmdc_automation/re_iding/scripts/re_id_tool.py:285]: the data object alternative identifier should be should be old_dms_id, not old data object identifier. this should be the original value for has_output of the omics_processing_set record. see https://github.com/microbiomedata/nmdc-schema/blob/abe2cf1ba29103bdc03d5e8463169e90ff387943/nmdc_schema/metab_id_refactor.py#L601C110-L601C120 which calls the original update_data_products function in Yuri's code.

**Review by aclum — CHANGES_REQUESTED** [review]
see comments about raw data object alternative_identifiers
https://github.com/microbiomedata/nmdc-schema/blob/abe2cf1ba29103bdc03d5e8463169e90ff387943/nmdc_schema/metab_id_refactor.py#L736

File: nmdc_automation/re_iding/scripts/re_id_tool.py [hunk 5]
```diff
@@ -92,6 +91,219 @@ def cli(ctx, target):
...
+    new_input_data_object_ids = []
+    for input_data_object in input_data_objects:
+        # mint new ID for the data object
+        legacy_id = input_data_object["id"]
+        new_id = api_client.minter("nmdc:DataObject")
+        new_input_data_object_ids.append(new_id)
+        updated_record_identifiers.append(("data_object_set", legacy_id, new_id))
+        # update the data object record
+        do_filter_criteria = {"id": legacy_id}
+        do_update_criteria = {"$set": {"id": new_id, "alternative_identifiers": [legacy_id]}}
```
[the remark this answers is not in the document]
Comment (mbthornton-lbl) [reply nmdc_automation/re_iding/scripts/re_id_tool.py:265]: I am using the `metabolomics.has_input` data objects, which I am assuming to be equal to `omics.has_output`
data objects

**Comment by mbthornton-lbl** [comment]
Add a check before each update stage to check if the object already has a valid ID

File: nmdc_automation/re_iding/scripts/re_id_tool.py [hunk 6]
```diff
@@ -92,6 +91,219 @@ def cli(ctx, target):
...
+            )
+        result = db_client["metabolomics_analysis_activity_set"].delete_one(
+            {"id": orphan_metabolomics_activity_record["id"]}
+            )
+        logging.info(f"Deleted {result.deleted_count} metabolomics_analysis_activity_set records")
+
+
+def _update_metabolomics_input_data_objects(input_data_objects, db_client, api_client, updated_record_identifiers):
+    """
+    Update the input data objects for a Metabolomics OmicsProcessing record.
```
Comment (aclum) [inline nmdc_automation/re_iding/scripts/re_id_tool.py:254]: Minor but function comment should say "Metabolomics Workflow Activity records". The inputs to OmicsProcessing is a biosample or processed sample so the description doesn't make sense.

File: nmdc_automation/re_iding/scripts/re_id_tool.py (hunk 4 above)
Comment (aclum) [inline nmdc_automation/re_iding/scripts/re_id_tool.py:285]: alternative identifiers should not be set for output data objects from the workflow execution activity. see line 732. https://github.com/microbiomedata/nmdc-schema/blob/abe2cf1ba29103bdc03d5e8463169e90ff387943/nmdc_schema/metab_id_refactor.py#L732
we are only keeping alt identifiers if they make sense in another outside system.

Comment (aclum) [reply nmdc_automation/re_iding/scripts/re_id_tool.py:285]: How do you want to handle that the first pass of this code incorrectly set generated_by, that needs to be unset ad hoc or in the code. ie napa data object id nmdc:dobj-11-fwgfyg40

**Review by aclum — CHANGES_REQUESTED** [review]
see comments on _update_metabolomics_output_data_objects function.

File: nmdc_automation/re_iding/scripts/re_id_tool.py (hunk 6 above)
Comment (mbthornton-lbl) [reply nmdc_automation/re_iding/scripts/re_id_tool.py]: https://github.com/microbiomedata/nmdc_automation/pull/116/commits/beebb0ae186d10adf9af0422bc228e7236426c18

File: nmdc_automation/re_iding/scripts/re_id_tool.py (hunk 4 above)
Comment (mbthornton-lbl) [reply nmdc_automation/re_iding/scripts/re_id_tool.py]: https://github.com/microbiomedata/nmdc_automation/pull/116/commits/beebb0ae186d10adf9af0422bc228e7236426c18

Comment (mbthornton-lbl) [reply nmdc_automation/re_iding/scripts/re_id_tool.py]: The data objects with `generated_by` were accidentally deleted

File: nmdc_automation/re_iding/scripts/re_id_tool.py [hunk 7]
```diff
@@ -92,6 +91,242 @@ def cli(ctx, target):
...
+    """
+    Update the calibration data object for a Metabolomics Workflow Activity record.
+    """
+    # mint new ID for the data object
+    legacy_id = calibration_data_object["id"]
+    new_id = api_client.minter("nmdc:DataObject")
+    updated_record_identifiers.append(("data_object_set", legacy_id, new_id))
+    # update the data object record
+    do_filter_criteria = {"id": legacy_id}
+    do_update_criteria = {"$set": {"id": new_id, "alternative_identifiers": [legacy_id]}}
```
Comment (aclum) [inline nmdc_automation/re_iding/scripts/re_id_tool.py:325]: remove updating alternative identifier for the calibration data_object_set records. This should only happen for data objects where are outputs of metabolomics or nom.

Comment (mbthornton-lbl) [reply nmdc_automation/re_iding/scripts/re_id_tool.py]: https://github.com/microbiomedata/nmdc_automation/pull/116/commits/a30883e042258efc7c5a087a303788e397c95945

**Review by aclum — CHANGES_REQUESTED** [review]
data objects that are values for has_calibration should not have alternative identifiers.

**Comment by mbthornton-lbl** [comment]
> data objects that are values for has_calibration should not have alternative identifiers.

Addressed here:
https://github.com/microbiomedata/nmdc_automation/commit/a30883e042258efc7c5a087a303788e397c95945

**Review by aclum — APPROVED** [review]

File: nmdc_automation/re_iding/base.py [hunk 8]
```diff
@@ -19,14 +19,20 @@
...
-                                                READ_BASED_TAXONOMY_ANALYSIS_ACTIVITY_SET,
-                                               check_for_single_omics_processing_record,
-                                               get_data_object_record_by_id,
-                                               get_omics_processing_id)
+from nmdc_automation.re_iding.db_utils import (
+    BIOSAMPLE_SET,
+    OMICS_PROCESSING_SET,
+    READS_QC_SET,
+    METAGENOME_ASSEMBLY_SET,
+    METATRANSCRIPTOME_ACTIVITY_SET,
```
Comment (aclum) [inline nmdc_automation/re_iding/base.py:27]: the metatranscriptome_activity_set collection contains records that are using annotations, we are deleting all of those so these should be deleted also. Perhaps have you have it is okay, in previous testing this entire collection ends up empty.

File: nmdc_automation/re_iding/base.py [hunk 9]
```diff
@@ -728,12 +762,79 @@ def _update_omics_processing_alternative_identifiers(omics_processing: nmdc.Omic
...
+def update_omics_output_data_object(
+        data_object: nmdc.DataObject, nmdc_omics_processing_id: str,
+        api_client: NmdcRuntimeApi, identifiers_map: dict=None) -> nmdc.DataObject:
+    """
+    Update the data object record with the new ID, and add the legacy ID to the alternate identifiers
+    """
+    updated_data_object = deepcopy(data_object)
+    # Check if we need to update the data object ID and set the alternative identifiers to the legacy ID
+    if not updated_data_object.id.startswith("nmdc:dobj-"):
+        updated_data_object.alternative_identifiers = [data_object.id]
```
Comment (aclum) [inline nmdc_automation/re_iding/base.py:774]: data object omics has_output alternative identifiers should only be set for nom,metabolomics. Proteomics omics has_output also have alternative identifiers but those are being added as entirely new records.

File: nmdc_automation/re_iding/base.py [hunk 10]
```diff
@@ -728,12 +762,79 @@ def _update_omics_processing_alternative_identifiers(omics_processing: nmdc.Omic
...
+    updated_metabolomics_analysis_activity.type = "nmdc:MetabolomicsAnalysisActivity"
+
+    # Check if we need to update the metabolomics analysis activity ID
+    if not updated_metabolomics_analysis_activity.id.startswith("nmdc:wfmb-"):
+        new_metabolomics_analysis_activity_id = get_new_nmdc_id(updated_metabolomics_analysis_activity, api_client, identifiers_map)
+        updated_metabolomics_analysis_activity.id = new_metabolomics_analysis_activity_id
+
+    updated_metabolomics_analysis_activity.was_informed_by = nmdc_omics_processing_id
+    updated_metabolomics_analysis_activity.has_input = [nmdc_input_data_object_id]
+    updated_metabolomics_analysis_activity.has_output = nmdc_output_data_object_ids
```
Comment (aclum) [inline nmdc_automation/re_iding/base.py:814]: has_output should be an array, not a string.

File: nmdc_automation/re_iding/base.py [hunk 11]
```diff
@@ -728,12 +762,79 @@ def _update_omics_processing_alternative_identifiers(omics_processing: nmdc.Omic
...
+                                 identifiers_map: dict=None) -> nmdc.NomAnalysisActivity:
+    """
+    Update the NOM analysis activity record with the new ID
+    """
+    updated_nom_analysis_activity = deepcopy(nom_analysis_activity)
+    # Ensure that type is set to nmdc:NomAnalysisActivity
+    updated_nom_analysis_activity.type = "nmdc:NomAnalysisActivity"
+
+    # Check if we need to update the NOM analysis activity ID
+    if not updated_nom_analysis_activity.id.startswith("nmdc:wfna-"):
```
Comment (aclum) [inline nmdc_automation/re_iding/base.py:832]: prefix is wrong, should be "nmdc:wfnom-"

File: nmdc_automation/re_iding/base.py [hunk 12]
```diff
@@ -728,12 +762,79 @@ def _update_omics_processing_alternative_identifiers(omics_processing: nmdc.Omic
...
+        new_nom_analysis_activity_id = get_new_nmdc_id(updated_nom_analysis_activity, api_client, identifiers_map)
+        updated_nom_analysis_activity.id = new_nom_analysis_activity_id
+
+    updated_nom_analysis_activity.was_informed_by = nmdc_omics_processing_id
+    updated_nom_analysis_activity.has_input = [nmdc_input_data_object_id]
+    updated_nom_analysis_activity.has_output = nmdc_output_data_object_ids
```
Comment (aclum) [inline nmdc_automation/re_iding/base.py:838]: has_output should be an array, not a string.

File: nmdc_automation/re_iding/scripts/re_id_tool.py [hunk 13]
```diff
@@ -157,10 +175,12 @@ def update_study(ctx, legacy_study_id, nmdc_study_id,  mongo_uri, identifiers_fi
     updates = {
         "study_set": {},
         "biosample_set": {},
-        "omics_processing_set": {}
+        "omics_processing_set": {},
+        "data_object_set": {},
+        "metabolomics_analysis_activity_set": {},
```
Comment (aclum) [inline nmdc_automation/re_iding/scripts/re_id_tool.py:180]: Should this list include nom_analysis_activity_set?

File: nmdc_automation/re_iding/base.py (hunk 12 above)
Comment (mbthornton-lbl) [reply nmdc_automation/re_iding/base.py:838]: The input argument `nmdc_output_data_object_ids` is a list.
```python
def update_nom_analysis_activity(nom_analysis_activity: nmdc.NomAnalysisActivity,
                                 nmdc_omics_processing_id: str, nmdc_input_data_object_id: str,
                                 nmdc_output_data_object_ids: List[str], api_client: NmdcRuntimeApi,
                                 identifiers_map: dict=None) -> nmdc.NomAnalysisActivity:
```

**Review by aclum — CHANGES_REQUESTED** [review]
See individual comments.

File: nmdc_automation/re_iding/scripts/re_id_tool.py (hunk 13 above)
Comment (mbthornton-lbl) [reply nmdc_automation/re_iding/scripts/re_id_tool.py:180]: Yes -

Comment (mbthornton-lbl) [reply nmdc_automation/re_iding/scripts/re_id_tool.py:180]: https://github.com/microbiomedata/nmdc_automation/pull/116/commits/54b01bc0144d7d1486bd471a49d007ff8d063af3

File: nmdc_automation/re_iding/base.py (hunk 11 above)
Comment (mbthornton-lbl) [reply nmdc_automation/re_iding/base.py]: https://github.com/microbiomedata/nmdc_automation/pull/116/commits/8d48f5cf35127ba6f231046d9f0ebd171b2573cc

File: nmdc_automation/re_iding/base.py (hunk 9 above)
Comment (mbthornton-lbl) [reply nmdc_automation/re_iding/base.py]: https://github.com/microbiomedata/nmdc_automation/pull/116/commits/8d48f5cf35127ba6f231046d9f0ebd171b2573cc

**Review by aclum — APPROVED** [review]

**Pull request merged**